nixos/modules/services/development/jupyterhub/default.nix

```diff
@@ -0,0 +1,108 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.jupyterhub;
+
+  jupyterhubConfig = pkgs.writeText "jupyterhub_config.py" ''
+    c.JupyterHub.bind_url = "http://${cfg.host}:${toString cfg.port}"
+
+    c.JupyterHub.authentication_class = "${cfg.authentication}"
+    c.JupyterHub.spawner_class = "${cfg.spawner}"
+
+    ${cfg.extraConfig}
+  '';
+
+  jupyterhub_env = pkgs.python3.withPackages (p: with p; cfg.packages);
+in {
+  meta.maintainers = with maintainers; [ costrouc ];
+
+  options.services.jupyterhub = {
+    enable = mkEnableOption "Jupyterhub development server";
+
+    authentication = mkOption {
+      type = types.str;
+      default = "jupyterhub.auth.PAMAuthenticator";
+      description = ''
+        Jupyterhub authentication to use
+      '';
+    };
+
+    spawner = mkOption {
+      type = types.str;
+      default = "systemdspawner.SystemdSpawner";
+      description = ''
+        Jupyterhub spawner to use
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.str;
+      default = "";
+      description = ''
+        Extra contents appended to the jupyterhub configuration
+      '';
+    };
+
+    packages = mkOption {
```

Comment:
It's unlikely that users need to change this so it may not need to be public. Regardless, consider calling it `jupyterhubEnv` where the default is `python3.withPackages(ps: with ps; [ jupyterhub jupyter-systemdspawner])`. This makes clearer that the packages are of the same Python packages set.

Reply:
Fixed. I have provided the `jupyterhubEnv` option. Having this option configurable will give a lot of flexibility on the hub which is required for more complex setups where we need to authentication with oauth or use kubernetes as a spawner. I have done my best to minimize the number of configuration options.